jormungandr/src/utils/fire_forget_scheduler.rs

```diff
@@ -0,0 +1,286 @@
+use futures::{Async, Future, Poll, Stream};
+use std::collections::VecDeque;
+use std::convert::Infallible;
+use std::time::Duration;
+use thiserror::Error;
+use tokio::sync::mpsc::{
+    self,
+    error::{RecvError, TrySendError},
+    Receiver, Sender,
+};
+use tokio::timer::delay_queue::{DelayQueue, Key};
+
+#[derive(Error, Debug)]
+pub enum Error {
+    #[error("failed to send a command: {0}")]
+    CommandSend(&'static str),
+    #[error("failed to receive a command")]
+    CommandReceive(#[from] RecvError),
+    #[error("command queue closed")]
+    CommandQueueClosed,
+    #[error("timer error")]
+    Timer(#[from] tokio::timer::Error),
+}
+
+impl<T> From<TrySendError<T>> for Error {
+    fn from(error: TrySendError<T>) -> Self {
+        let cause = if error.is_closed() {
+            "channel closed"
+        } else if error.is_full() {
+            "no available capacity"
+        } else {
+            "unknown channel error"
+        };
+        Error::CommandSend(cause)
+    }
+}
+
+/// Schedule for fire-forget tasks
+///
+/// Each task has an ID (TID) and can be launched onto a worker identified with an ID (WID).
+/// Launching is defined by a closure (LAUNCHER), which also accepts additional data (DATA).
+/// Multiple instances of task with same TID, but different WID may be running in parallel.
+/// Launcher must be quick and non-blocking.
+///
+/// When task is scheduled, it's queued. When it's finally launched, a timeout is started.
+/// Until it runs out or task is declared complete it's considered running and it consumes parallel
+/// task execution limits. Finishing with timeout doesn't make task completed and other queued
+/// instances with same TID, but different WID may be run. Timed out instances are considered
+/// failed and aren't rescheduled. Finishing by declaration task completion cancels all task
+/// instances with the same TID.
```

Comment:
I'm not sure I get the semantics of this. If we want to run competing pulls against different peers, would that mean the pull tasks receive the same TID, but have different data?

Reply:
They all should have same TID, but because they can be run in parallel, they need different WID. The most sensible WID in this case is `NodeID`, because the node will be processing the request. The DATA is not shared between tasks or workers, each scheduling has its own DATA, it can contain anything.